# RAG Answer Quality Evaluation

Retrieval evaluation tells us whether the right chunk was found. But it doesn't tell us whether the final answer is good. A chunk might not rank first but still provide enough context for a good answer.

We use **LLM-as-a-judge** to evaluate answer quality:
1. Sample 200 questions from ground truth
2. Run RAG for each question
3. Ask GPT-4o-mini to judge the answer as `RELEVANT`, `PARTLY_RELEVANT`, or `NON_RELEVANT`
4. Aggregate the results

### Why LLM-as-a-judge?

Manually evaluating 200 answers would take hours. LLM-as-a-judge is a common and effective approach for automated evaluation of open-ended text generation. The judge LLM reads the question and generated answer and classifies relevance.

**Limitation:** The judge LLM may have its own biases. In practice, human evaluation is the gold standard but LLM-as-a-judge is a good proxy for rapid iteration.

### Search pipeline

All configurations use the same retrieval pipeline selected in `04_evaluation-ret-exp.ipynb` and `05_evaluation-ret-exp.ipynb`:
- **Hybrid search** — combines vector search and full text search with RRF
- **Cross-encoder reranking** — `cross-encoder/ms-marco-MiniLM-L-6-v2` reranks the top 20 hybrid search results and returns the top 10
- **Embedding model** — `pritamdeka/S-PubMedBert-MS-MARCO` — best performing biomedical model from `04_evaluation-ret-exp.ipynb`

### Configurations compared

We test two dimensions:
- **LLM model** — `gpt-4o-mini` (cheaper, faster) vs `gpt-5.6-terra` (more capable, more expensive)
- **Prompt** — default prompt (detailed instructions) vs concise prompt (brief, citation-focused)

#### Step 1. Load ground truth questions and subsample them

In [ ]:
from app.ragbase import RAGBaseHistory, INSTRUCTIONS
from sentence_transformers import SentenceTransformer
import json
from tqdm.auto import tqdm
from openai import OpenAI
from pipelines.evaluation import calculate_cost


c:\Users\marta\Documents\projects\literature-assistant-llm\.venv\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [ ]:
import pandas as pd

df_question = pd.read_csv("../data/ground_truth.csv")
df_sample = df_question.sample(n=200, random_state=1)

sample = df_sample.to_dict(orient="records")

In [3]:
import psycopg

conn = psycopg.connect(
    "postgresql://user:pswd@localhost:5432/papers"
)

print("Connected!")

Connected!


In [9]:
prompt_template_rag_eval = """
You are an expert evaluator for a RAG system.
Your task is to analyze the relevance of the generated answer to the given question.
Based on the relevance of the generated answer, classify it
as NON_RELEVANT, PARTLY_RELEVANT, or RELEVANT.

Here is the data for evaluation:

Question: {question}
Generated Answer: {answer_llm}

Please analyze the content and context of the generated answer in relation to the question
and provide your evaluation in parsable JSON without using code blocks:

{{
  "Relevance": "NON_RELEVANT" | "PARTLY_RELEVANT" | "RELEVANT",
  "Explanation": "[Provide a brief explanation for your evaluation]"
}}
""".strip()

### Step 2. Define configurations to test

In [10]:
concise_prompt = """
You are a scientific assistant. Answer the question based ONLY on the provided context.
Be concise and precise. Cite the source (author, year) for each claim.
If the answer is not in the context, say "I don't know."
""".strip()

In [ ]:
configurations = [
    {
        "name": "gpt-4o-mini, default prompt",
        "llm_model": "gpt-4o-mini",
        "instructions": INSTRUCTIONS  # current default in ragbase
    },
    {
        "name": "gpt-4o-mini, concise prompt",
        "llm_model": "gpt-4o-mini",
        "instructions": concise_prompt
    },
    {
        "name": "gpt-5.6-terra, default prompt",
        "llm_model": "gpt-5.6-terra",
        "instructions": INSTRUCTIONS
    },
        {
        "name": "gpt-5.6-terra, concise prompt",
        "llm_model": "gpt-5.6-terra",
        "instructions": concise_prompt
    }
]

In [ ]:
# pricing per million tokens (as of 2026)
PRICING = {
    "gpt-4o-mini": {"input": 0.15, "output": 0.60},
    "gpt-5.6-terra": {"input": 2.0, "output": 12.00}
}


### Step 3. Evaluate different rag configurations

In [13]:
openai_client = OpenAI()
embedding_model = SentenceTransformer("pritamdeka/S-PubMedBert-MS-MARCO")

all_results = []

for config in configurations:
    print(f"\nEvaluating: {config['name']}")
    
    assistant = RAGBaseHistory(
        conn=conn,
        model=embedding_model,
        llm_client=openai_client,
        llm_model=config["llm_model"],
        instructions=config["instructions"]
    )

    evaluations = []
    total_cost = 0.0
    
    for record in tqdm(sample):
        question = record["question"]

        assistant.clear_history()

        answer = assistant.rag(question)
        answer_llm = answer.answer
        usage_llm = answer.usage

        rag_cost = calculate_cost(usage_llm, config["llm_model"])
        total_cost += rag_cost

        prompt = prompt_template_rag_eval.format(
            question=question,
            answer_llm=answer_llm
        )

        eval_response = openai_client.responses.create(
            model="gpt-4o-mini",  # always use mini for judging
            input=[{"role": "user", "content": prompt}]
        )

        try:
            evaluation = json.loads(eval_response.output_text)
        except json.JSONDecodeError:
            evaluation = {"Relevance": "NON_RELEVANT", "Explanation": "Parse error"}

        evaluations.append({
            "record": record,
            "answer_llm": answer_llm,
            "usage": usage_llm,
            "evaluation": evaluation,
            "config": config["name"]
        })

    df_eval = pd.DataFrame(evaluations)
    df_eval["question"] = df_eval.record.apply(lambda d: d["question"])
    df_eval["relevance"] = df_eval.evaluation.apply(lambda d: d["Relevance"])
    df_eval["config"] = config["name"]
    
    relevant_pct = (df_eval["relevance"] == "RELEVANT").mean()
    partly_pct = (df_eval["relevance"] == "PARTLY_RELEVANT").mean()
    non_pct = (df_eval["relevance"] == "NON_RELEVANT").mean()
    
    all_results.append({
        "config": config["name"],
        "relevant": relevant_pct,
        "partly_relevant": partly_pct,
        "non_relevant": non_pct,
        "total_cost_usd": round(total_cost, 4),  
        "cost_per_question": round(total_cost / len(sample), 6)  
    })
    
    print(f"Relevant: {relevant_pct:.1%}, Partly: {partly_pct:.1%}, Non: {non_pct:.1%}")
    
    # save each config results
    df_eval.to_csv(f"data/rag-eval-{config['name'].replace(' ', '-')}.csv", index=False)

# final comparison
df_results = pd.DataFrame(all_results)
print("\nFinal comparison:")
print(df_results)

Loading weights: 100%|██████████| 199/199 [00:00<00:00, 23104.32it/s]



Evaluating: gpt-4o-mini, default prompt


100%|██████████| 200/200 [19:55<00:00,  5.98s/it]


Relevant: 94.0%, Partly: 4.5%, Non: 1.5%

Evaluating: gpt-4o-mini, concise prompt


100%|██████████| 200/200 [16:41<00:00,  5.01s/it]


Relevant: 93.5%, Partly: 3.0%, Non: 3.5%

Evaluating: gpt-4o, default prompt


100%|██████████| 200/200 [19:25<00:00,  5.83s/it]


Relevant: 92.5%, Partly: 5.0%, Non: 2.5%

Evaluating: gpt-4o, concise prompt


100%|██████████| 200/200 [17:00<00:00,  5.10s/it]

Relevant: 89.0%, Partly: 7.5%, Non: 3.5%

Final comparison:
                        config  relevant  partly_relevant  non_relevant  \
0  gpt-4o-mini, default prompt     0.940            0.045         0.015   
1  gpt-4o-mini, concise prompt     0.935            0.030         0.035   
2       gpt-4o, default prompt     0.925            0.050         0.025   
3       gpt-4o, concise prompt     0.890            0.075         0.035   

   total_cost_usd  cost_per_question  
0          0.1061           0.000530  
1          0.0974           0.000487  
2          1.5948           0.007974  
3          1.4276           0.007138  


In [ ]:
df_results.to_csv("../data/rag-eval.csv", index=False)

## Conclusion

Based on the RAG answer quality evaluation, **GPT-4o-mini with the default prompt** was selected as the final configuration:

| Configuration | Relevant | Partly relevant | Non-relevant | Cost/200 questions |
|--------------|----------|----------------|--------------|-------------------|
| **gpt-4o-mini, default prompt** ✅ | **94.0%** | 4.5% | 1.5% | $0.11 |
| gpt-4o-mini, concise prompt | 93.5% | 3.0% | 3.5% | $0.10 |
| gpt-5.6-terra, default prompt | 92.5% | 5.0% | 2.5% | $1.59 |
| gpt-5.6-terra, concise prompt | 89.0% | 7.5% | 3.5% | $1.43 |

**Key findings:**

- **GPT-4o-mini performs as well as GPT-5.6-terra** — the more expensive model does not provide meaningful quality improvement (~1-2% difference)
- **GPT-4o-mini is ~15x cheaper** — $0.11 vs $1.59 for 200 questions
- **Default prompt outperforms concise prompt** — detailed instructions with encouragement to use multiple sources produces more relevant answers
- **Overall quality is high** — 94% relevant answers despite the challenging domain of highly similar scientific papers

**Selected configuration:**
- LLM model: `gpt-4o-mini`
- Prompt: default `INSTRUCTIONS` from `ragbase.py`

This configuration is used in the final RAG pipeline (`RAGBaseHistory` in `ragbase.py`) together with hybrid search + cross-encoder reranking selected in `05_evaluation-ret-exp2.ipynb`.